src/expr/node_manager.cpp

```diff
@@ -686,8 +696,270 @@ Node NodeManager::mkAbstractValue(const TypeNode& type) {
   return n;
 }
 
+void NodeManager::newPolymorphicFunction(TNode n){
+  Assert ( n.getType(false).isFunction() );
+  Assert ( !isPolymorphicFunctionInstance(n) );
+  Assert ( !isPolymorphicFunction(n) );
+  d_polymorphicFunction[n] = n;
+}
+
+bool NodeManager::isPolymorphicFunction(TNode n){
+  return d_polymorphicFunction.find(n) != d_polymorphicFunction.end();
+}
+
+bool NodeManager::isPolymorphicFunctionInstance(TNode n){
+  return d_instanceFunction.find(n) != d_instanceFunction.end();
+}
+
+TNode NodeManager::getPolymorphicFunction(TNode n){
+  std::hash_map<Node, Node, NodeHashFunction>::const_iterator i = d_instanceFunction.find(n);
+  Assert ( n.getType(false).isFunction() );
+  if (i != d_instanceFunction.end()) {
+    return (*i).second;
+  } else {
+    return TNode::null();
+  }
+}
+
+TNode NodeManager::getPolymorphicFunctionFromPolymorphicInstance(TNode n){
+  std::hash_map<Node, Node, NodeHashFunction>::const_iterator i = d_polymorphicFunction.find(n);
+  Assert ( n.getType(false).isFunction() );
+  if (i != d_polymorphicFunction.end()) {
+    return (*i).second;
+  } else {
+    return TNode::null();
+  }
+}
+
+bool isCloseSchemaVar(NodeManager& nm, TypeNode ty){
+
+  if (ty.getNumChildren() == 0) {
+    /* variable case */
+    return !nm.isPolymorphicTypeVarSchema(ty);
+
+  } else {
+    /* arity n */
+
+    for(size_t i = 0, len = ty.getNumChildren(); i < len; ++i) {
+      if(!(isCloseSchemaVar(nm,ty[i]))) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  return false;
+
+}
+
+/**
+   The matching is done without Subtyping,
+   t2 must be ground
+*/
+bool NodeManager::matchPolymorphicType(TypeNode t1,
+                                      TypeNode t2,
+                                      std::hash_map<TypeNode, TypeNode, TypeNode::HashFunction>& subst)
+  throw(TypeCheckingExceptionPrivate){
+
+  if(isPolymorphicTypeVarSchema(t2)){
+    throw TypeCheckingExceptionPrivate(Node::null(),"No inference is done, you should add an (as term ty) for specifying the return type.");
```

Comment:
`No inference is done` or `No inference [will ever] be carried out`? 
The former would mean that inference will be done, but this error was triggered before that phase.
The latter means that there is no inference pass.